ddl/table_lock.go

```diff
@@ -0,0 +1,230 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ddl
+
+import (
+	"github.com/pingcap/errors"
+	"github.com/pingcap/parser/model"
+	"github.com/pingcap/tidb/infoschema"
+	"github.com/pingcap/tidb/meta"
+)
+
+func onLockTables(t *meta.Meta, job *model.Job) (ver int64, err error) {
+	arg := &lockTablesArg{}
+	if err := job.DecodeArgs(arg); err != nil {
+		// Invalid arguments, cancel this job.
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+
+	// Unlock table first.
+	if arg.IndexOfUnlock < len(arg.UnlockTables) {
+		return unlockTables(t, job, arg)
+	}
+
+	// Check table locked by other, this can be only checked at the first time.
+	if arg.IndexOfLock == 0 {
+		for i, tl := range arg.LockTables {
+			job.SchemaID = tl.SchemaID
+			job.TableID = tl.TableID
+			tbInfo, err := getTableInfoAndCancelFaultJob(t, job, job.SchemaID)
+			if err != nil {
+				return ver, err
+			}
+			err = checkTableLocked(tbInfo, arg.LockTables[i].Tp, arg.SessionInfo)
+			if err != nil {
+				// If any request table was locked by other session, just cancel this job.
+				// No need to rolling back the unlocked tables, MySQL will release the lock first
+				// and block if the request table was locked by other.
+				job.State = model.JobStateCancelled
+				return ver, errors.Trace(err)
+			}
+		}
+	}
+
+	// Lock tables.
+	if arg.IndexOfLock < len(arg.LockTables) {
+		job.SchemaID = arg.LockTables[arg.IndexOfLock].SchemaID
+		job.TableID = arg.LockTables[arg.IndexOfLock].TableID
+		var tbInfo *model.TableInfo
+		tbInfo, err = getTableInfoAndCancelFaultJob(t, job, job.SchemaID)
+		if err != nil {
+			return ver, err
+		}
+		err = lockTable(tbInfo, arg.IndexOfLock, arg)
+		if err != nil {
+			job.State = model.JobStateCancelled
+			return ver, err
+		}
+
+		switch tbInfo.Lock.State {
+		case model.TableLockStateNone:
+			// none -> pre_lock
+			tbInfo.Lock.State = model.TableLockStatePreLock
+			tbInfo.Lock.TS = t.StartTS
+			ver, err = updateVersionAndTableInfo(t, job, tbInfo, true)
```

Comment:
Do we need to update the job's `TableID`? The TableID always use the first ID. If the lock statement has multiple locks，it will some thing wrong in checking schema validate.

Reply:
`job.TableID` and `job.SchemaID` already updated in line#59.